File: sdk/applocksdk/src/main/kotlin/com/applocksdk/LockEvent.kt
package com.applocksdk

object LockEventType {
    const val LOCK_TRIGGERED = "LOCK_TRIGGERED"
    const val UNLOCK_SUCCESS = "UNLOCK_SUCCESS"
    const val UNLOCK_FAILED = "UNLOCK_FAILED"
    const val MAX_ATTEMPTS_REACHED = "MAX_ATTEMPTS_REACHED"
    const val LOCK_BY_TIMEOUT = "LOCK_BY_TIMEOUT"
    const val BIOMETRIC_RECOGNITION_FAILED = "BIOMETRIC_RECOGNITION_FAILED"
}

data class LockEvent(
    val appId: String,
    val deviceId: String,
    val event: String,
    val method: String,
    val attempt: Int,
    val timestamp: String,
    val sdkVersion: String,
    val osVersion: String,
    val manufacturer: String,
    val model: String,
    val sdkInt: Int,
    val language: String,
    val appVersion: String,
)

File: server/src/models/LockEvent.js
const mongoose = require("mongoose");

const lockEventSchema = new mongoose.Schema({
  appId: { type: String, required: true },
  deviceId: { type: String, required: true },
  event: { type: String, required: true },
  method: { type: String, required: true },
  attempt: { type: Number, required: true },
  timestamp: { type: Date, required: true },
  sdkVersion: { type: String, required: true },
  osVersion: { type: String, required: true },
  manufacturer: { type: String, required: true },
  model: { type: String, required: true },
  sdkInt: { type: Number, required: true },
  language: { type: String, required: true },
  appVersion: { type: String, required: true },
});

// Analytics queries by appId + time.
lockEventSchema.index({ appId: 1, timestamp: -1 });
// Detect suspicious devices - failed count per device.
lockEventSchema.index({ appId: 1, deviceId: 1, event: 1, timestamp: -1 });
// Auto-delete events older than 90 days.
lockEventSchema.index({ timestamp: 1 }, { expireAfterSeconds: 7776000 });

module.exports = mongoose.model("LockEvent", lockEventSchema, "lock_events");
